src/librustdoc/test.rs

```diff
@@ -346,6 +391,17 @@ fn run_test(test: &str, cratename: &str, filename: &FileName, line: usize,
     };
     cmd.env(var, &newpath);
 
+    if persist_executable {
+        let mut target_path = target_path.unwrap();
+        target_path.push("path.txt");
+
+        std::fs::write(
+            target_path,
+            newpath.as_os_str().to_str().unwrap().as_bytes(),
```

Comment:
This seems a little awkward, because it's not always the same environment variable that's being set here. A better option would be to make this an environment file, and make the file output be the string "`{var}={newpath}`", to at least show that this should be setting something other than `PATH` ([on non-Windows platforms, at least](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/dynamic_lib/struct.DynamicLibrary.html#method.envvar)). (With these file contents, i'd probably also name it something like `test.env`, to make it more obvious what it's there for.)

One situation where this breaks down is if the doctests are being built for a different target than the host. It looks like the chosen environment variable will always be stuck to what the *host* expects, but if you're expecting to run the doctests on another target (say, some embedded platform), then that may be different. However, in this situation it's also possible that the dylib path extension may not be necessary. Still, i think the above suggestion should be good enough.

Reply:
Huh, I never considered taking in the path, sounds like a much better idea! 👌

Reply:
Ah, okay, yeah fair enough :)

Reply:
I'm pretty sure we don't need any of this `newpath` stuff now that we're not compiling doctests with `prefer-dynamic` since #54939. I don't think we need a `path.txt` or `test.env` file at all.

Reply:
I'm guessing you may have been using a version of rustdoc that didn't have #54939 applied yet. It'll be a good idea to rebase this PR and try again. Anyway I've submitted  #56514 to see if anything breaks.